File: src/personal_data_warehouse_photos/scanner.py
"""Apple Photos library scanner.

Reads a snapshot of the library's ``Photos.sqlite`` (never the live file: WAL
churn from Photos.app would tear reads) and resolves each non-trashed asset to
the original PhotoKit resource(s) to export: the primary photo/video, plus a
Live Photo's original paired-video component. The scanner intentionally does
not inspect ``originals/``: Optimize Mac Storage makes that directory an
incomplete cache, not an authoritative view of the library.
"""

from __future__ import annotations

import hashlib
import json
import os
import sqlite3
import threading
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any

# Core Data stores timestamps as seconds since 2001-01-01T00:00:00Z.
COCOA_EPOCH_UNIX_OFFSET = 978307200

# Photos stores "no GPS" as -180.0 in both latitude and longitude.
GPS_UNSET_SENTINEL = -180.0

# ZASSET.ZKINDSUBTYPE for a Live Photo still.
KIND_SUBTYPE_LIVE_PHOTO = 2

_UTI_MIME_TYPES = {
    "public.heic": "image/heic",
    "public.heif": "image/heif",
    "public.jpeg": "image/jpeg",
    "public.png": "image/png",
    "public.tiff": "image/tiff",
    "org.webmproject.webp": "image/webp",
    "com.compuserve.gif": "image/gif",
    "com.adobe.raw-image": "image/x-adobe-dng",
    "com.apple.quicktime-movie": "video/quicktime",
    "public.mpeg-4": "video/mp4",
}

_EXTENSION_MIME_TYPES = {
    ".heic": "image/heic",
    ".heif": "image/heif",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".png": "image/png",
    ".tif": "image/tiff",
    ".tiff": "image/tiff",
    ".webp": "image/webp",
    ".gif": "image/gif",
    ".dng": "image/x-adobe-dng",
    ".mov": "video/quicktime",
    ".mp4": "video/mp4",
}


class ApplePhotosSchemaError(RuntimeError):
    pass


PROBE_TIMEOUT_SECONDS = 15.0


def _probe_openable(path: Path, timeout_seconds: float = PROBE_TIMEOUT_SECONDS) -> None:
    """Fail fast when opening the Photos library would hang.

    Unlike ~/Library/Messages (where a launchd process without Full Disk
    Access gets an immediate EPERM), open(2) on the Photos-library files can
    BLOCK indefinitely inside macOS TCC. A hung run holds the uploader lock
    forever, never writes a heartbeat exit code, and reads as healthy — the
    exact silent-failure mode the heartbeat work exists to prevent. Probe the
    open in a daemon thread and convert a stall into the same loud
    PermissionError the other uploaders raise.
    """
    outcome: list[BaseException | None] = []

    def attempt() -> None:
        try:
            handle = os.open(path, os.O_RDONLY)
            os.close(handle)
            outcome.append(None)
        except BaseException as exc:  # noqa: BLE001 - re-raised below with guidance
            outcome.append(exc)

    thread = threading.Thread(target=attempt, name="photos-open-probe", daemon=True)
    thread.start()
    thread.join(timeout_seconds)
    if not outcome:
        raise PermissionError(
            f"Opening {path} blocked for {timeout_seconds:.0f}s — macOS TCC is stalling this "
            "process. Grant Full Disk Access to the launching executable chain "
            "(see the Apple Photos section of AGENTS.md) and retry."
        )
    if isinstance(outcome[0], BaseException):
        raise PermissionError(
            f"Could not open Apple Photos store at {path}. Grant Full Disk Access to the "
            "launching executable chain and retry."
        ) from outcome[0]


@dataclass(frozen=True)
class PhotoFileCandidate:
    """One original PhotoKit resource claim."""

    native_id: str  # the still asset's ZUUID for BOTH roles (this attaches the .mov)
    role: str  # original | live_video
    asset_kind: str  # image | video (of the parent asset)
    filename: str
    extension: str
    mime_type: str
    expected_size_bytes: int
    width: int
    height: int
    captured_at: str  # local wall-clock ISO, "" when unknown
    capture_tz_offset: str  # "+HH:MM" / "-HH:MM" / ""
    camera_make: str
    camera_model: str
    apple_record: dict[str, Any]

    @property
    def fingerprint(self) -> str:
        payload = json.dumps(
            {"version": 2, "role": self.role, "apple_record": self.apple_record},
            sort_keys=True,
            separators=(",", ":"),
        ).encode("utf-8")
        return f"photokit-v2|{hashlib.sha256(payload).hexdigest()}"

    @property
    def state_id(self) -> str:
        return f"{self.native_id}|{self.role}"


def snapshot_photos_store(library_path: Path | str, destination_dir: Path | str) -> Path:
    """Copy Photos.sqlite via the sqlite backup API (repo precedent: apple-messages)."""
    source_path = Path(library_path).expanduser() / "database" / "Photos.sqlite"
    destination = Path(destination_dir) / "Photos.sqlite"
    _probe_openable(source_path)
    source_uri = source_path.resolve().as_uri() + "?mode=ro"
    try:
        source = sqlite3.connect(source_uri, uri=True)
    except sqlite3.Error as exc:
        raise PermissionError(
            f"Could not open Apple Photos store at {source_path}. "
            "Grant Full Disk Access to the launching executable chain and retry."
        ) from exc
    try:
        target = sqlite3.connect(destination)
        try:
            source.backup(target)
        finally:
            target.close()
    finally:
        source.close()
    return destination


def scan_photo_file_candidates(
    snapshot_path: Path | str,
) -> list[PhotoFileCandidate]:
    """Newest-first original-resource candidates for every non-trashed asset."""
    connection = sqlite3.connect(Path(snapshot_path))
    connection.row_factory = sqlite3.Row
    try:
        tables = {
            str(row["name"])
            for row in connection.execute("SELECT name FROM sqlite_master WHERE type = 'table'")
        }
        required = {"ZASSET", "ZADDITIONALASSETATTRIBUTES", "ZEXTENDEDATTRIBUTES"}
        missing = required - tables
        if missing:
            raise ApplePhotosSchemaError(
                f"Unsupported Apple Photos schema: missing {', '.join(sorted(missing))}"
            )
        rows = connection.execute(
            """
            SELECT
                a.ZUUID AS uuid,
                a.ZDIRECTORY AS directory,
                a.ZFILENAME AS filename,
                a.ZKIND AS kind,
                a.ZKINDSUBTYPE AS kind_subtype,
                a.ZUNIFORMTYPEIDENTIFIER AS uti,
                a.ZDATECREATED AS date_created,
                a.ZADDEDDATE AS added_date,
                a.ZMODIFICATIONDATE AS modification_date,
                a.ZWIDTH AS width,
                a.ZHEIGHT AS height,
                a.ZLATITUDE AS latitude,
                a.ZLONGITUDE AS longitude,
                a.ZFAVORITE AS favorite,
                a.ZHIDDEN AS hidden,
                a.ZADJUSTMENTSSTATE AS adjustments_state,
                aa.ZORIGINALFILENAME AS original_filename,
                aa.ZORIGINALFILESIZE AS original_file_size,
                aa.ZTIMEZONEOFFSET AS timezone_offset,
                aa.ZINFERREDTIMEZONEOFFSET AS inferred_timezone_offset,
                aa.ZTIMEZONENAME AS timezone_name,
                aa.ZEXIFTIMESTAMPSTRING AS exif_timestamp_string,
                ea.ZCAMERAMAKE AS camera_make,
                ea.ZCAMERAMODEL AS camera_model,
                ea.ZLENSMODEL AS lens_model
            FROM ZASSET a
            LEFT JOIN ZADDITIONALASSETATTRIBUTES aa ON aa.ZASSET = a.Z_PK
            LEFT JOIN ZEXTENDEDATTRIBUTES ea ON ea.ZASSET = a.Z_PK
            WHERE a.ZTRASHEDSTATE = 0
            ORDER BY a.ZDATECREATED DESC
            """
        ).fetchall()
    finally:
        connection.close()

    candidates: list[PhotoFileCandidate] = []
    for row in rows:
        candidates.extend(_candidates_for_asset(row))
    return candidates


def _candidates_for_asset(row: sqlite3.Row) -> list[PhotoFileCandidate]:
    uuid = str(row["uuid"] or "")
    filename = str(row["filename"] or "")
    if not uuid or not filename:
        return []
    kind = int(row["kind"] or 0)
    asset_kind = "video" if kind == 1 else "image"

    tz_seconds = _first_int(row["timezone_offset"], row["inferred_timezone_offset"])
    captured_at, tz_offset = _wall_clock(row["date_created"], tz_seconds)
    record = _apple_record(row)

    original_filename = str(row["original_filename"] or "") or filename
    extension = _suffix(original_filename)
    original = _file_candidate(
        row,
        native_id=uuid,
        role="original",
        asset_kind=asset_kind,
        filename=original_filename,
        extension=extension,
        mime_type=_mime_type(str(row["uti"] or ""), extension),
        captured_at=captured_at,
        tz_offset=tz_offset,
        record=record,
    )
    candidates = [original]

    if asset_kind == "image" and int(row["kind_subtype"] or 0) == KIND_SUBTYPE_LIVE_PHOTO:
        # PhotoKit exposes the original motion component as a paired-video
        # resource. It is uploaded under the SAME native id, which lets the
        # identity layer attach it to the still's asset.
        live_name = f"{Path(original_filename).stem}.MOV"
        candidates.append(
            _file_candidate(
                row,
                native_id=uuid,
                role="live_video",
                asset_kind=asset_kind,
                filename=live_name,
                extension=".mov",
                mime_type="video/quicktime",
                captured_at=captured_at,
                tz_offset=tz_offset,
                record=record,
            )
        )
    return candidates


def _file_candidate(
    row: sqlite3.Row,
    *,
    native_id: str,
    role: str,
    asset_kind: str,
    filename: str,
    extension: str,
    mime_type: str,
    captured_at: str,
    tz_offset: str,
    record: dict[str, Any],
) -> PhotoFileCandidate:
    width = int(row["width"] or 0) if role == "original" else 0
    height = int(row["height"] or 0) if role == "original" else 0
    return PhotoFileCandidate(
        native_id=native_id,
        role=role,
        asset_kind=asset_kind,
        filename=filename,
        extension=extension,
        mime_type=mime_type,
        expected_size_bytes=int(row["original_file_size"] or 0) if role == "original" else 0,
        width=width,
        height=height,
        captured_at=captured_at,
        capture_tz_offset=tz_offset,
        camera_make=str(row["camera_make"] or ""),
        camera_model=str(row["camera_model"] or ""),
        apple_record=record,
    )


def _apple_record(row: sqlite3.Row) -> dict[str, Any]:
    """The archival raw payload: everything the scanner read, losslessly."""
    record: dict[str, Any] = {
        "uuid": str(row["uuid"] or ""),
        "kind": int(row["kind"] or 0),
        "kind_subtype": int(row["kind_subtype"] or 0),
        "uti": str(row["uti"] or ""),
        "directory": str(row["directory"] or ""),
        "filename": str(row["filename"] or ""),
        "original_filename": str(row["original_filename"] or ""),
        "date_created": _iso_utc(row["date_created"]),
        "added_date": _iso_utc(row["added_date"]),
        "modification_date": _iso_utc(row["modification_date"]),
        "timezone_name": str(row["timezone_name"] or ""),
        "timezone_offset_seconds": _first_int(row["timezone_offset"], row["inferred_timezone_offset"]),
        "exif_timestamp_string": str(row["exif_timestamp_string"] or ""),
        "width": int(row["width"] or 0),
        "height": int(row["height"] or 0),
        "original_file_size": int(row["original_file_size"] or 0),
        "favorite": int(row["favorite"] or 0),
        "hidden": int(row["hidden"] or 0),
        "adjustments_state": int(row["adjustments_state"] or 0),
        "camera_make": str(row["camera_make"] or ""),
        "camera_model": str(row["camera_model"] or ""),
        "lens_model": str(row["lens_model"] or ""),
    }
    latitude = _gps_value(row["latitude"])
    longitude = _gps_value(row["longitude"])
    if latitude is not None and longitude is not None:
        record["latitude"] = latitude
        record["longitude"] = longitude
    return record


def _gps_value(value: Any) -> float | None:
    if value is None:
        return None
    number = float(value)
    if number == GPS_UNSET_SENTINEL:
        return None
    return number


def _first_int(*values: Any) -> int:
    for value in values:
        if value is not None:
            return int(value)
    return 0


def _cocoa_datetime(value: Any) -> datetime | None:
    if value is None:
        return None
    return datetime.fromtimestamp(float(value) + COCOA_EPOCH_UNIX_OFFSET, tz=UTC)


def _iso_utc(value: Any) -> str:
    moment = _cocoa_datetime(value)
    return moment.isoformat() if moment else ""


def _wall_clock(date_created: Any, tz_seconds: int) -> tuple[str, str]:
    """(local wall-clock ISO without offset, "+HH:MM" offset) for the capture."""
    moment = _cocoa_datetime(date_created)
    if moment is None:
        return "", ""
    local = moment + timedelta(seconds=tz_seconds)
    sign = "+" if tz_seconds >= 0 else "-"
    magnitude = abs(tz_seconds)
    offset = f"{sign}{magnitude // 3600:02d}:{(magnitude % 3600) // 60:02d}"
    return local.strftime("%Y-%m-%dT%H:%M:%S"), offset


def _suffix(filename: str) -> str:
    return Path(filename).suffix.lower()


def _mime_type(uti: str, extension: str) -> str:
    if uti in _UTI_MIME_TYPES:
        return _UTI_MIME_TYPES[uti]
    if extension in _EXTENSION_MIME_TYPES:
        return _EXTENSION_MIME_TYPES[extension]
    return "application/octet-stream"

File: app/internal/uploaders/photos/scanner.go
// Package photos is the Go port of the Apple Photos uploader
// (personal_data_warehouse_photos): it snapshots the library's Photos.sqlite,
// resolves every user-library asset to the original PhotoKit resource(s) to
// export, exports each one through the native helper app with iCloud access,
// and uploads the complete bytes plus the cross-source metadata envelope
// through the app's photo ingest endpoints.
//
// Every wire payload, fingerprint and dedup key is byte-identical to what the
// Python uploader produced: the warehouse dedups on sha256 of canonical JSON
// and the state file keys on the fingerprint, so a one-byte drift would
// re-upload the whole library on the first Go run. testdata/golden.jsonl is
// the Python output the tests pin this against.
package photos

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

// CocoaEpochUnixOffset is 2001-01-01T00:00:00Z as Unix seconds; Core Data
// stores timestamps relative to it.
const CocoaEpochUnixOffset = 978307200

// GPSUnsetSentinel is how Photos stores "no GPS" in both latitude and longitude.
const GPSUnsetSentinel = -180.0

// KindSubtypeLivePhoto is ZASSET.ZKINDSUBTYPE for a Live Photo still.
const KindSubtypeLivePhoto = 2

// ProbeTimeout bounds how long opening the Photos store may block inside
// macOS TCC before the run is failed loudly.
const ProbeTimeout = 15 * time.Second

var utiMimeTypes = map[string]string{
	"public.heic":               "image/heic",
	"public.heif":               "image/heif",
	"public.jpeg":               "image/jpeg",
	"public.png":                "image/png",
	"public.tiff":               "image/tiff",
	"org.webmproject.webp":      "image/webp",
	"com.compuserve.gif":        "image/gif",
	"com.adobe.raw-image":       "image/x-adobe-dng",
	"com.apple.quicktime-movie": "video/quicktime",
	"public.mpeg-4":             "video/mp4",
}

var extensionMimeTypes = map[string]string{
	".heic": "image/heic",
	".heif": "image/heif",
	".jpg":  "image/jpeg",
	".jpeg": "image/jpeg",
	".png":  "image/png",
	".tif":  "image/tiff",
	".tiff": "image/tiff",
	".webp": "image/webp",
	".gif":  "image/gif",
	".dng":  "image/x-adobe-dng",
	".mov":  "video/quicktime",
	".mp4":  "video/mp4",
}

// SchemaError reports a Photos.sqlite without the tables the scanner reads.
type SchemaError struct{ Missing []string }

func (e *SchemaError) Error() string {
	return "Unsupported Apple Photos schema: missing " + strings.Join(e.Missing, ", ")
}

// Candidate is one original PhotoKit resource claim.
type Candidate struct {
	NativeID          string // the still asset's ZUUID for BOTH roles (this attaches the .mov)
	Role              string // original | live_video
	AssetKind         string // image | video (of the parent asset)
	Filename          string
	Extension         string
	MimeType          string
	ExpectedSizeBytes int64
	Width             int64
	Height            int64
	CapturedAt        string // local wall-clock ISO, "" when unknown
	CaptureTZOffset   string // "+HH:MM" / "-HH:MM" / ""
	CameraMake        string
	CameraModel       string
	AppleRecord       map[string]any
}

// Fingerprint is the stable Photos-metadata fingerprint the state file keys
// on: "photokit-v2|" + sha256 of the canonical {version, role, apple_record}.
func (c Candidate) Fingerprint() string {
	return "photokit-v2|" + common.JSONSHA256(map[string]any{
		"version":      int64(2),
		"role":         c.Role,
		"apple_record": c.AppleRecord,
	})
}

// StateID is the upload_state source_id: "<native_id>|<role>".
func (c Candidate) StateID() string {
	return c.NativeID + "|" + c.Role
}

// probeOpenable fails fast when opening the Photos library would hang.
//
// Unlike ~/Library/Messages (where a launchd process without Full Disk
// Access gets an immediate EPERM), open(2) on the Photos-library files can
// BLOCK indefinitely inside macOS TCC. A hung run holds the uploader lock
// forever, never writes a heartbeat exit code, and reads as healthy. Probe
// the open in a goroutine and convert a stall into the same loud permission
// error the other uploaders raise.
func probeOpenable(path string, timeout time.Duration) error {
	outcome := make(chan error, 1)
	go func() {
		file, err := os.Open(path)
		if err == nil {
			file.Close()
		}
		outcome <- err
	}()
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	select {
	case err := <-outcome:
		if err != nil {
			return &common.PermissionError{Path: path, Err: err}
		}
		return nil
	case <-ctx.Done():
		return &common.PermissionError{Path: path, Err: fmt.Errorf(
			"opening %s blocked for %.0fs; macOS TCC is stalling this process (see the Apple Photos section of AGENTS.md)",
			path, timeout.Seconds())}
	}
}

// SnapshotStore copies the library's Photos.sqlite into destinationDir
// (never reading the live file directly: WAL churn from Photos.app would
// tear reads) and returns the snapshot path.
func SnapshotStore(libraryPath, destinationDir string) (string, error) {
	source := filepath.Join(common.ExpandUser(libraryPath), "database", "Photos.sqlite")
	destination := filepath.Join(destinationDir, "Photos.sqlite")
	if err := probeOpenable(source, ProbeTimeout); err != nil {
		return "", err
	}
	if err := common.SnapshotSQLite(source, destination); err != nil {
		return "", err
	}
	return destination, nil
}

const candidateQuery = `
SELECT
    a.ZUUID AS uuid,
    a.ZDIRECTORY AS directory,
    a.ZFILENAME AS filename,
    a.ZKIND AS kind,
    a.ZKINDSUBTYPE AS kind_subtype,
    a.ZUNIFORMTYPEIDENTIFIER AS uti,
    a.ZDATECREATED AS date_created,
    a.ZADDEDDATE AS added_date,
    a.ZMODIFICATIONDATE AS modification_date,
    a.ZWIDTH AS width,
    a.ZHEIGHT AS height,
    a.ZLATITUDE AS latitude,
    a.ZLONGITUDE AS longitude,
    a.ZFAVORITE AS favorite,
    a.ZHIDDEN AS hidden,
    a.ZADJUSTMENTSSTATE AS adjustments_state,
    aa.ZORIGINALFILENAME AS original_filename,
    aa.ZORIGINALFILESIZE AS original_file_size,
    aa.ZTIMEZONEOFFSET AS timezone_offset,
    aa.ZINFERREDTIMEZONEOFFSET AS inferred_timezone_offset,
    aa.ZTIMEZONENAME AS timezone_name,
    aa.ZEXIFTIMESTAMPSTRING AS exif_timestamp_string,
    ea.ZCAMERAMAKE AS camera_make,
    ea.ZCAMERAMODEL AS camera_model,
    ea.ZLENSMODEL AS lens_model
FROM ZASSET a
LEFT JOIN ZADDITIONALASSETATTRIBUTES aa ON aa.ZASSET = a.Z_PK
LEFT JOIN ZEXTENDEDATTRIBUTES ea ON ea.ZASSET = a.Z_PK
WHERE a.ZTRASHEDSTATE = 0
  -- Bundle-scoped rows are transient syndicated/shared records stored in
  -- Photos.sqlite but are not user-library PHAssets. PhotoKit cannot fetch
  -- or export them.
  AND COALESCE(a.ZBUNDLESCOPE, 0) = 0
ORDER BY a.ZDATECREATED DESC
`

// Scan returns newest-first original-resource candidates for every
// non-trashed, user-library asset in a Photos.sqlite snapshot.
func Scan(snapshotPath string) ([]Candidate, error) {
	db, err := common.OpenSQLite(snapshotPath)
	if err != nil {
		return nil, err
	}
	defer db.Close()
	tables, err := common.TableNames(db)
	if err != nil {
		return nil, err
	}
	var missing []string
	for _, required := range []string{"ZADDITIONALASSETATTRIBUTES", "ZASSET", "ZEXTENDEDATTRIBUTES"} {
		if !tables[required] {
			missing = append(missing, required)
		}
	}
	if len(missing) > 0 {
		return nil, &SchemaError{Missing: missing}
	}
	rows, err := common.Query(db, candidateQuery)
	if err != nil {
		return nil, err
	}
	var candidates []Candidate
	for _, row := range rows {
		candidates = append(candidates, candidatesForAsset(row)...)
	}
	return candidates, nil
}

func candidatesForAsset(row common.Row) []Candidate {
	uuid := strOr(row.Get("uuid"))
	filename := strOr(row.Get("filename"))
	if uuid == "" || filename == "" {
		return nil
	}
	assetKind := "image"
	if common.ToInt(row.Get("kind")) == 1 {
		assetKind = "video"
	}
	tzSeconds := firstInt(row.Get("timezone_offset"), row.Get("inferred_timezone_offset"))
	capturedAt, tzOffset := wallClock(row.Get("date_created"), tzSeconds)
	record := appleRecord(row)

	originalFilename := strOr(row.Get("original_filename"))
	if originalFilename == "" {
		originalFilename = filename
	}
	extension := suffix(originalFilename)
	candidates := []Candidate{fileCandidate(row, uuid, "original", assetKind, originalFilename, extension,
		mimeType(strOr(row.Get("uti")), extension), capturedAt, tzOffset, record)}

	if assetKind == "image" && common.ToInt(row.Get("kind_subtype")) == KindSubtypeLivePhoto {
		// PhotoKit exposes the original motion component as a paired-video
		// resource. It is uploaded under the SAME native id, which lets the
		// identity layer attach it to the still's asset.
		liveName := stem(originalFilename) + ".MOV"
		candidates = append(candidates, fileCandidate(row, uuid, "live_video", assetKind, liveName, ".mov",
			"video/quicktime", capturedAt, tzOffset, record))
	}
	return candidates
}

func fileCandidate(row common.Row, nativeID, role, assetKind, filename, extension, mime, capturedAt, tzOffset string, record map[string]any) Candidate {
	var width, height, expected int64
	if role == "original" {
		width = common.ToInt(row.Get("width"))
		height = common.ToInt(row.Get("height"))
		expected = common.ToInt(row.Get("original_file_size"))
	}
	return Candidate{
		NativeID:          nativeID,
		Role:              role,
		AssetKind:         assetKind,
		Filename:          filename,
		Extension:         extension,
		MimeType:          mime,
		ExpectedSizeBytes: expected,
		Width:             width,
		Height:            height,
		CapturedAt:        capturedAt,
		CaptureTZOffset:   tzOffset,
		CameraMake:        strOr(row.Get("camera_make")),
		CameraModel:       strOr(row.Get("camera_model")),
		AppleRecord:       record,
	}
}

// appleRecord is the archival raw payload: everything the scanner read.
func appleRecord(row common.Row) map[string]any {
	record := map[string]any{
		"uuid":                    strOr(row.Get("uuid")),
		"kind":                    common.ToInt(row.Get("kind")),
		"kind_subtype":            common.ToInt(row.Get("kind_subtype")),
		"uti":                     strOr(row.Get("uti")),
		"directory":               strOr(row.Get("directory")),
		"filename":                strOr(row.Get("filename")),
		"original_filename":       strOr(row.Get("original_filename")),
		"date_created":            isoUTC(row.Get("date_created")),
		"added_date":              isoUTC(row.Get("added_date")),
		"modification_date":       isoUTC(row.Get("modification_date")),
		"timezone_name":           strOr(row.Get("timezone_name")),
		"timezone_offset_seconds": firstInt(row.Get("timezone_offset"), row.Get("inferred_timezone_offset")),
		"exif_timestamp_string":   strOr(row.Get("exif_timestamp_string")),
		"width":                   common.ToInt(row.Get("width")),
		"height":                  common.ToInt(row.Get("height")),
		"original_file_size":      common.ToInt(row.Get("original_file_size")),
		"favorite":                common.ToInt(row.Get("favorite")),
		"hidden":                  common.ToInt(row.Get("hidden")),
		"adjustments_state":       common.ToInt(row.Get("adjustments_state")),
		"camera_make":             strOr(row.Get("camera_make")),
		"camera_model":            strOr(row.Get("camera_model")),
		"lens_model":              strOr(row.Get("lens_model")),
	}
	latitude, latOK := gpsValue(row.Get("latitude"))
	longitude, lonOK := gpsValue(row.Get("longitude"))
	if latOK && lonOK {
		record["latitude"] = latitude
		record["longitude"] = longitude
	}
	return record
}

func gpsValue(value any) (float64, bool) {
	if value == nil {
		return 0, false
	}
	number := common.ToFloat(value)
	if number == GPSUnsetSentinel {
		return 0, false
	}
	return number, true
}

// firstInt mirrors _first_int: int() of the first non-NULL value, else 0.
func firstInt(values ...any) int64 {
	for _, value := range values {
		if value != nil {
			return common.ToInt(value)
		}
	}
	return 0
}

// strOr mirrors str(value or ""): NULL, 0 and "" all read as "".
func strOr(value any) string {
	switch v := value.(type) {
	case nil:
		return ""
	case int64:
		if v == 0 {
			return ""
		}
	case float64:
		if v == 0 {
			return ""
		}
	}
	return common.PyStr(value)
}

func cocoaTime(value any) (time.Time, bool) {
	if value == nil {
		return time.Time{}, false
	}
	return common.UnixFromFloat(common.ToFloat(value) + CocoaEpochUnixOffset), true
}

func isoUTC(value any) string {
	moment, ok := cocoaTime(value)
	if !ok {
		return ""
	}
	return common.ISOFormat(moment)
}

// wallClock returns (local wall-clock ISO without offset, "+HH:MM") for the
// capture.
func wallClock(dateCreated any, tzSeconds int64) (string, string) {
	moment, ok := cocoaTime(dateCreated)
	if !ok {
		return "", ""
	}
	local := moment.Add(time.Duration(tzSeconds) * time.Second)
	sign := "+"
	if tzSeconds < 0 {
		sign = "-"
	}
	magnitude := tzSeconds
	if magnitude < 0 {
		magnitude = -magnitude
	}
	offset := fmt.Sprintf("%s%02d:%02d", sign, magnitude/3600, (magnitude%3600)/60)
	return local.Format("2006-01-02T15:04:05"), offset
}

// suffix mirrors Path(filename).suffix.lower().
func suffix(filename string) string {
	return strings.ToLower(suffixRaw(filepath.Base(filename)))
}

// stem mirrors Path(filename).stem.
func stem(filename string) string {
	base := filepath.Base(filename)
	return strings.TrimSuffix(base, suffixRaw(base))
}

// suffixRaw is pathlib's suffix: the last dot onward, unless it is the
// leading character (a dotfile has no suffix).
func suffixRaw(base string) string {
	i := strings.LastIndex(base, ".")
	if i <= 0 {
		return ""
	}
	return base[i:]
}

func mimeType(uti, extension string) string {
	if mime, ok := utiMimeTypes[uti]; ok {
		return mime
	}
	if mime, ok := extensionMimeTypes[extension]; ok {
		return mime
	}
	return "application/octet-stream"
}
